pilot/cmd/pilot-agent/main.go

```diff
@@ -413,12 +422,72 @@ var (
 
 			log.Infof("PilotSAN %#v", pilotSAN)
 
-			envoyProxy := envoy.NewProxy(proxyConfig, role.ServiceNode(), proxyLogLevel, proxyComponentLogLevel, pilotSAN, role.IPAddresses, dnsRefreshRate, opts)
-			agent := envoy.NewAgent(envoyProxy, envoy.DefaultRetry, features.TerminationDrainDuration())
-			watcher := envoy.NewWatcher(tlsCertsToWatch, agent.ConfigCh())
+			// TODO(nmittler): Once the new Envoy manager has baked, get rid of the variable and the old code.
+			if useNewEnvoyManager.Get() {
+				// Create a function for running the Envoy restart manager.
+				runEnvoyManager := func() restart.Manager {
+					mgr := restart.New(restart.Config{
+						MaxRestartAttempts:       uint(envoy.DefaultRetry.MaxRetries),
+						InitialRestartInterval:   envoy.DefaultRetry.InitialInterval,
```

Comment:
this is maybe outside the scope of this PR but related -- I am starting to think this whole restarting on envoy crash with backoff etc is just shooting ourselves in the foot for no reason. Consider a case where Envoy is going OOM.
* Envoy dies, but kubernetes still reports things are good and 2/2 ready. Repeat a few times.
* At somepoint due to backoff, envoy is dead for 2 minutes. The readiness probe won't fail for at least a minute, so there is 1min where traffic is coming and we are not even attempting to handle it.
* Meanwhile, the operator has no clue what is going on, because Kubernetes doesn't know envoy is crashing

Alternatively, if we exit when envoy crashes:
* The readiness probe fails immediately
* The user can clearly see envoy is crashing
* Kubernetes handles backoff for us
* We don't have to handle the complexity.

We do still need hot restarts, but I think that is orthogonal to restarting failed Envoy's

Reply:
That's a good point.  Seems reasonable to me.

@kyessenov @costinm @rshriram  WDYT?

Reply:
The restart was addressing races in hot-restart protocol and not permanent failures (like bad bootstrap). We can't really distinguish the two, and we know that hot restart has race conditions.

Reply:
If we're using SDS can we simply turn off restarts altogether? In the termination logic, we abort all processes created by the manager, which *should* resolve hot-restart races IIUC.

Reply:
Hot restart race happens when a starting envoy talks to a running envoy, and there's some mis-understanding over the protocol, which usually causes the new envoy to crash. If I remember, we found at least 3 segfaults falling in this category, and still not clear if any are left. Usually, another attempt helps to resolve since the running envoy can re-negotiate handling over the sockets.

Reply:
@kyessenov my understanding is if we intentionally trigger a hot restart (certs), there is no traffic loss. But if Envoy crashes, its no longer a hot restart -- its just a normal restart, with normal traffic lost. In this case why do we need pilot-agent to restart envoy, rather than just having kubernetes do it? I think if kubernetes does it it will immediately get marked as unready as well which has added benefit.

Reply:
I think leaving restarts to k8s makes sense.  If we built hot restart into envoy.Instance, we could effectively get rid of this manager altogether. 

@kyessenov are we missing other cases for why pilot-agent is managing the restarts? VMs perhaps?

Reply:
`systemd` manages restarts on VMs I think?